runtime/jvmti/jvmtiHook.c

```diff
@@ -2776,6 +2785,41 @@ jvmtiHookCompilingEnd(J9HookInterface** hook, UDATA eventNum, void* eventData, v
 
 #endif /* INTERP_NATIVE_SUPPORT */
 
+static void 
+jvmtiHookSampledObjectAlloc(J9HookInterface** hook, UDATA eventNum, void* eventData, void* userData)
+{
+	J9JVMTIEnv * j9env = userData;
+	jvmtiEventSampledObjectAlloc callback = j9env->callbacks.SampledObjectAlloc;
+
+	Trc_JVMTI_jvmtiHookSampledObjectAlloc_Entry();
+
+	ENSURE_EVENT_PHASE_LIVE(jvmtiHookSampledObjectAlloc, j9env);
+
+	if (NULL != callback) {
+		J9SampledObjectAllocateEvent *data = eventData;
+		J9VMThread *currentThread = data->currentThread;
+		jthread threadRef = NULL;
+		UDATA hadVMAccess = 0;
+		UDATA javaOffloadOldState = 0;
+
+		if (prepareForEvent(j9env, currentThread, currentThread, JVMTI_EVENT_SAMPLED_OBJECT_ALLOC, &threadRef, &hadVMAccess, TRUE, 2, &javaOffloadOldState)) {
+			j9object_t *objectRef = (j9object_t*) currentThread->arg0EA;
+			j9object_t *classRef = (j9object_t*) (currentThread->arg0EA - 1);
+			J9Class *clazz = NULL;
+			J9InternalVMFunctions const * const vmFuncs = currentThread->javaVM->internalVMFunctions;
+
+			*objectRef = data->object;
+			clazz = J9OBJECT_CLAZZ(currentThread, data->object);
+			*classRef = J9VM_J9CLASS_TO_HEAPCLASS(clazz);
+			vmFuncs->internalExitVMToJNI(currentThread);
+			callback((jvmtiEnv *) j9env, (JNIEnv *) currentThread, threadRef, (jobject) objectRef, (jclass) classRef, (jlong) data->size);
+			vmFuncs->internalEnterVMFromJNI(currentThread);
```

Comment:
data->object should be updated here (with VM access!) with the contents of objectRef in case there's more than one hook listener.

Reply:
Fixed. Assuming this is to protect the heap object over GC point.